### PR TITLE
BUG-116378 – credential

### PR DESCRIPTION
**Pull request by gregito**
BUG-116378 – credential

### CODE REVIEW

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Base.java [hunk 1]
```diff
@@ -7,33 +7,39 @@
...
+@ApiModel(description = "Base class/JSON which contains all the default fields for Credential related requests/responses",
+        subTypes = {CredentialV4Request.class, CredentialV4Response.class})
+public abstract class CredentialV4Base implements JsonEntity {
+
     @Size(max = 100, min = 5, message = "The length of the credential's name has to be in range of 5 to 100")
     @Pattern(regexp = "(^[a-z][-a-z0-9]*[a-z0-9]$)",
             message = "The name of the credential can only contain lowercase alphanumeric characters and hyphens and has start with an alphanumeric character")
     @NotNull
-    @ApiModelProperty(value = ModelDescriptions.NAME, required = true)
+    @ApiModelProperty(value = ModelDescriptions.NAME, required = true, allowableValues = "any text which meets with the format constraints")
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Base.java]: allowableValues is not required, only for exact values (enums, range etc...)

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Base.java [hunk 2]
```diff
@@ -7,33 +7,39 @@
...
     private Map<String, Object> parameters = new HashMap<>();
 
     @Size(max = 1000)
-    @ApiModelProperty(ModelDescriptions.DESCRIPTION)
+    @ApiModelProperty(value = ModelDescriptions.DESCRIPTION, allowableValues = "any text with the maximum length of 1000 character")
     private String description;
 
-    @ApiModelProperty(ModelDescriptions.TOPOLOGY_ID)
+    @ApiModelProperty(value = ModelDescriptions.TOPOLOGY_ID,
+            allowableValues = "any whole number from -9,223,372,036,854,775,808 to 9,223,372,036,854,775,807", notes = "Most likely a positive number")
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Base.java]: invalid allowable values

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/responses/InteractiveCredentialResponse.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+
+@ApiModel(value = "InteractiveCredential", description = "Contains values from an Azure interactive login attempt")
+public class InteractiveCredentialResponse {
+
+    @ApiModelProperty(value = "The user code what has to be used for the sign-in process on the Azure portal", example = "B8ZT2QD4K")
+    private String userCode;
+
+    @ApiModelProperty(value = "The url provided by Azure where the user have to use the given user code to sign in",
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/responses/InteractiveCredentialResponse.java]: use ModelDescriptions class

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/responses/CredentialV4Response.java [hunk 4]
```diff
@@ -1,31 +1,37 @@
...
+    @Valid
+    @ApiModelProperty(value = "workspace", allowableValues = "any valid WorkspaceResourceResponse which refers to the workspace of the given credential")
     private WorkspaceResourceResponse workspace;
 
-    @ApiModelProperty(value = ModelDescriptions.GOV_CLOUD_FLAG)
+    @ApiModelProperty(value = ModelDescriptions.GOV_CLOUD_FLAG, allowableValues = "true of false")
     private Boolean govCloud;
 
-    @JsonProperty("public")
+    @JsonProperty(value = "public")
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/responses/CredentialV4Response.java]: we dont use anymore publicInAccount

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Base.java [hunk 5]
```diff
@@ -7,33 +7,39 @@
...
+    @ApiModelProperty(value = ModelDescriptions.NAME, required = true, allowableValues = "any text which meets with the format constraints")
     private String name;
 
     @NotNull
-    @ApiModelProperty(value = ModelDescriptions.CLOUD_PLATFORM, required = true)
+    @ApiModelProperty(value = ModelDescriptions.CLOUD_PLATFORM, required = true, allowableValues = "any text")
     private String cloudPlatform;
 
     @ApiModelProperty(CredentialModelDescription.PARAMETERS)
     private Map<String, Object> parameters = new HashMap<>();
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Base.java]: change to provider specific class (Aws, Azure, GCP, OpenStack)

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Base.java (hunk 2 above)
Comment (gregito) [reply core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Base.java]: why?

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v1/CredentialEndpoint.java [hunk 6]
```diff
@@ -14,8 +14,8 @@
 import javax.ws.rs.Produces;
 import javax.ws.rs.core.MediaType;
 
-import com.sequenceiq.cloudbreak.api.model.CredentialRequest;
-import com.sequenceiq.cloudbreak.api.model.CredentialResponse;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.credentials.requests.CredentialV4Request;
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v1/CredentialEndpoint.java]: remove v1 endpoint related classes

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Endpoint.java [hunk 7]
```diff
@@ -27,78 +28,79 @@
...
+            nickname = "getCredentialInWorkspace", httpMethod = "GET")
+    CredentialV4Response get(@PathParam("workspaceId") Long workspaceId, @PathParam("name") String name);
 
     @POST
     @Path("")
     @Produces(MediaType.APPLICATION_JSON)
     @ApiOperation(value = CredentialOpDescription.CREATE_IN_WORKSPACE, produces = ContentType.JSON, notes = Notes.CREDENTIAL_NOTES,
-            nickname = "createCredentialInWorkspace")
-    CredentialResponse createInWorkspace(@PathParam("workspaceId") Long workspaceId, @Valid CredentialRequest request);
+            nickname = "createCredentialInWorkspace", httpMethod = "POST", consumes = MediaType.APPLICATION_JSON)
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Endpoint.java]: Consumes is duplicated. We annotate the class with this media type

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Endpoint.java [hunk 8]
```diff
@@ -27,78 +28,79 @@
...
+    CredentialV4Response put(@PathParam("workspaceId") Long workspaceId, @Valid CredentialV4Request credentialRequest);
 
     @POST
     @Path("interactivelogin")
     @Produces(MediaType.APPLICATION_JSON)
     @ApiOperation(value = CredentialOpDescription.INTERACTIVE_LOGIN, produces = ContentType.JSON, notes = Notes.CREDENTIAL_NOTES,
-            nickname = "interactiveLoginCredentialInWorkspace")
-    Map<String, String> interactiveLogin(@PathParam("workspaceId") Long workspaceId, @Valid CredentialRequest credentialRequest);
+            nickname = "interactiveLoginCredentialInWorkspace", httpMethod = "POST", consumes = MediaType.APPLICATION_JSON)
+    InteractiveCredentialResponse interactiveLogin(@PathParam("workspaceId") Long workspaceId, @Valid CredentialV4Request credentialRequest);
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Endpoint.java]: please use V4 every resonse or request class.

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Endpoint.java [hunk 9]
```diff
@@ -27,78 +28,79 @@
...
     CredentialPrerequisites getPrerequisitesForCloudPlatform(@PathParam("workspaceId") Long workspaceId, @PathParam("cloudPlatform") String platform,
             @QueryParam("deploymentAddress") String deploymentAddress);
 
     @GET
     @Path("codegrantflow/init")
     @ApiOperation(value = CredentialOpDescription.INIT_CODE_GRANT_FLOW, notes = Notes.CREDENTIAL_NOTES,
-            nickname = "initCodeGrantFlowBasedCredentialInWorkspace")
-    Response initCodeGrantFlow(@PathParam("workspaceId") Long workspaceId, @Valid CredentialRequest credentialRequest);
+            nickname = "initCodeGrantFlowBasedCredentialInWorkspace", httpMethod = "GET", consumes = MediaType.APPLICATION_JSON)
+    Response initCodeGrantFlow(@PathParam("workspaceId") Long workspaceId, @Valid CredentialV4Request credentialRequest);
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Endpoint.java]: is this a query or update or create?

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Endpoint.java [hunk 10]
```diff
@@ -27,78 +28,79 @@
...
 
     @GET
     @Path("codegrantflow/init/{name}")
     @ApiOperation(value = CredentialOpDescription.INIT_CODE_GRANT_FLOW_ON_EXISTING, notes = Notes.CREDENTIAL_NOTES,
-            nickname = "initCodeGrantFlowOnExistingCredentialInWorkspace")
+            nickname = "initCodeGrantFlowOnExistingCredentialInWorkspace", httpMethod = "GET")
     Response initCodeGrantFlowOnExisting(@PathParam("workspaceId") Long workspaceId, @PathParam("name") String name);
 
     @GET
     @Path("codegrantflow/authorization/{cloudPlatform}")
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Endpoint.java]: CodeGrantFlow maybe a new endpoint?

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/aws/RoleBasedCredentialParameters.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+@ApiModel
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class RoleBasedCredentialParameters implements Mappable {
+
+    @ApiModelProperty(required = true, example = "arn:aws:iam::981628461338:role/example-role")
+    private String roleArn;
+
+    @ApiModelProperty
+    private String selector = AwsSelectorType.ROLE_BASED.getName();
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/aws/RoleBasedCredentialParameters.java]: Enum is usable for type in JSON. Jackson is mapping automatically the strings to correct enum.

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/cumulus/CumulusYarnCredentialParameters.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+
+@ApiModel
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class CumulusYarnCredentialParameters implements CredentialV4Parameters {
+
+    @ApiModelProperty(required = true)
+    private String cumulusAmbariPassword;
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/cumulus/CumulusYarnCredentialParameters.java]: cumulus is unnecessary in field names. Are we use Cumulus with Ambari only?

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/gcp/GcpCredentialParameters.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+import com.sequenceiq.cloudbreak.api.endpoint.v4.credentials.providers.CloudPlatform;
+
+import io.swagger.annotations.ApiModel;
+
+@ApiModel
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class GcpCredentialParameters implements CredentialV4Parameters {
+
+    private P12Parameters p12Parameters;
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/gcp/GcpCredentialParameters.java]: p12 and json is enough for field names

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/gcp/JsonParameters.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+
+@ApiModel
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class JsonParameters implements Mappable {
+
+    @ApiModelProperty(required = true)
+    private String credentialJson;
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/gcp/JsonParameters.java]: is this contains a json?

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/mock/MockCredentialParameters.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.credentials.parameters.CredentialV4Parameters;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.credentials.providers.CloudPlatform;
+
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+
+@ApiModel()
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/mock/MockCredentialParameters.java]: () is unnecessary

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/openstack/KeystoneV2Parameters.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+import com.sequenceiq.cloudbreak.api.endpoint.v4.credentials.parameters.Mappable;
+
+import io.swagger.annotations.ApiModelProperty;
+
+public class KeystoneV2Parameters implements Mappable {
+
+    @ApiModelProperty(required = true)
+    private String tenantName;
+
+    @ApiModelProperty(required = true, allowableValues = "cb-keystone-v2")
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/openstack/KeystoneV2Parameters.java]: I think this field unnecessary if one value is allowable

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/yarn/YarnCredentialParameters.java [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+@ApiModel
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class YarnCredentialParameters implements CredentialV4Parameters {
+
+    @ApiModelProperty(required = true)
+    private String cumulusAmbariUser;
+
+    @ApiModelProperty(required = true)
+    private String yarnEndpoint;
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/yarn/YarnCredentialParameters.java]: yarn is unnecessary in field names

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/responses/InteractiveCredentialResponse.java [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+import static com.sequenceiq.cloudbreak.doc.ModelDescriptions.EnvironmentRequestModelDescription.INTERACTIVE_LOGIN_CREDENTIAL_USER_CODE;
+import static com.sequenceiq.cloudbreak.doc.ModelDescriptions.EnvironmentRequestModelDescription.INTERACTIVE_LOGIN_CREDENTIAL_VERIFICATION_URL;
+
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+
+@ApiModel(value = "InteractiveCredential", description = "Contains values from an Azure interactive login attempt")
+public class InteractiveCredentialResponse {
+
+    @ApiModelProperty(value = INTERACTIVE_LOGIN_CREDENTIAL_USER_CODE, example = "B8ZT2QD4K")
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/responses/InteractiveCredentialResponse.java]: is B8ZT2QD4K a valid userCode or random?

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/userprofile/responses/UserProfileV4Response.java [hunk 19]
```diff
@@ -18,11 +21,13 @@
 
     private String tenant;
 
-    public CredentialResponse getCredential() {
+    private Map<String, Object> uiProperties = new HashMap<>();
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/userprofile/responses/UserProfileV4Response.java]: uiProperties are removed by Adam

File: core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/credentials/MapToAzureCredentialParametersConverter.java [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+
+import java.util.Map;
+
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.api.endpoint.v4.credentials.parameters.azure.AzureCredentialParameters;
+import com.sequenceiq.cloudbreak.converter.AbstractConversionServiceAwareConverter;
+
+@Component
+public class MapToAzureCredentialParametersConverter extends AbstractConversionServiceAwareConverter<Map<String, Object>, AzureCredentialParameters> {
```
Comment (topolyai5) [inline core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/credentials/MapToAzureCredentialParametersConverter.java]: I think it is changeable to something like `new Json(map).getValue(AzureCredentialParameters.class)`

File: core/src/main/java/com/sequenceiq/cloudbreak/service/credential/CredentialPropertyCollector.java [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+import com.sequenceiq.cloudbreak.api.endpoint.v4.credentials.providers.CloudPlatform;
+
+@Component
+public class CredentialPropertyCollector {
+
+    @Inject
+    private CredentialPropertyPropagator credentialPropertyPropagator;
+
+    public Map<String, Object> propertyMap(CredentialV4Base cred) {
+        CredentialV4Parameters alma = credentialPropertyPropagator.propagateCredentialProperty(cred)
```
Comment (topolyai5) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/credential/CredentialPropertyCollector.java]: alma?

Comment (gregito) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/credential/CredentialPropertyCollector.java]: 😅

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/responses/InteractiveCredentialResponse.java (hunk 18 above)
Comment (gregito) [reply core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/responses/InteractiveCredentialResponse.java]: a modified valid one

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/gcp/JsonParameters.java (hunk 14 above)
Comment (gregito) [reply core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/gcp/JsonParameters.java]: yes

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Base.java [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+    @ApiModelProperty(value = CredentialModelDescription.YARN_PARAMETERS, hidden = true)
+    private MockCredentialParameters mock;
+
+    @Size(max = 1000)
+    @ApiModelProperty(ModelDescriptions.DESCRIPTION)
+    private String description;
+
+    @ApiModelProperty(value = ModelDescriptions.TOPOLOGY_ID, allowableValues = "range[-9223372036854775808, 9223372036854775807]",
+            notes = "Most likely a positive number")
+    private Long topologyId;
```
Comment (doktoric) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Base.java]: do we need this ?

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Endpoint.java [hunk 23]
```diff
@@ -27,78 +28,79 @@
...
     @Path("")
     @Produces(MediaType.APPLICATION_JSON)
     @ApiOperation(value = CredentialOpDescription.PUT_IN_WORKSPACE, produces = ContentType.JSON, notes = Notes.CREDENTIAL_NOTES,
-            nickname = "putCredentialInWorkspace")
-    CredentialResponse putInWorkspace(@PathParam("workspaceId") Long workspaceId, @Valid CredentialRequest credentialRequest);
+            nickname = "putCredentialInWorkspace", httpMethod = "PUT")
+    CredentialV4Response put(@PathParam("workspaceId") Long workspaceId, @Valid CredentialV4Request credentialRequest);
 
     @POST
     @Path("interactivelogin")
```
Comment (doktoric) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Endpoint.java]: interactive_login

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Endpoint.java [hunk 24]
```diff
@@ -27,78 +28,79 @@
...
     @ApiOperation(value = CredentialOpDescription.INIT_CODE_GRANT_FLOW, notes = Notes.CREDENTIAL_NOTES,
-            nickname = "initCodeGrantFlowBasedCredentialInWorkspace")
-    Response initCodeGrantFlow(@PathParam("workspaceId") Long workspaceId, @Valid CredentialRequest credentialRequest);
+            nickname = "codeGrantFlowBasedCredentialInWorkspace", httpMethod = "GET")
+    Response initCodeGrantFlow(@PathParam("workspaceId") Long workspaceId, @Valid CredentialV4Request credentialRequest);
 
-    @GET
-    @Path("codegrantflow/init/{name}")
+    @PUT
+    @Path("codegrantflow/{name}")
```
Comment (doktoric) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Endpoint.java]: code_grant_flow

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/aws/AwsCredentialParameters.java [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+
+    public RoleBasedCredentialParameters getRoleBasedCredentialParameters() {
+        return roleBasedCredentialParameters;
+    }
+
+    public void setRoleBasedCredentialParameters(RoleBasedCredentialParameters roleBasedCredentialParameters) {
+        this.roleBasedCredentialParameters = roleBasedCredentialParameters;
+    }
+
+    public Boolean getGovCloud() {
```
Comment (doktoric) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/aws/AwsCredentialParameters.java]: boolean is better. what do you think ?

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/aws/KeyBasedCredentialParameters.java [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+
+    @NotNull
+    @ApiModelProperty(required = true, example = "ASIBJ34QYCJ1IBLK24KA")
+    private String accessKey;
+
+    @NotNull
+    @ApiModelProperty(required = true, example = "Ratk5cM9edxGuN6jdGb/8Jf621ZuTVGkoO14GPwN")
+    private String secretKey;
+
+    @ApiModelProperty
```
Comment (doktoric) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/aws/KeyBasedCredentialParameters.java]: defaultvalue

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/aws/RoleBasedCredentialParameters.java [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+@ApiModel
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class RoleBasedCredentialParameters implements Mappable {
+
+    @NotNull
+    @ApiModelProperty(required = true, example = "arn:aws:iam::981628461338:role/example-role")
+    private String roleArn;
+
+    @ApiModelProperty
```
Comment (doktoric) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/aws/RoleBasedCredentialParameters.java]: add default value

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/gcp/P12Parameters.java [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+    @NotNull
+    @ApiModelProperty(required = true, example = "serviceaccountemailaddress@example.com")
+    private String serviceAccountId;
+
+    @NotNull
+    @ApiModelProperty(required = true)
+    private String serviceAccountPrivateKey;
+
+    @ApiModelProperty
+    private GcpSelectorType selector = GcpSelectorType.P12;
```
Comment (doktoric) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/gcp/P12Parameters.java]: default value

File: core-api/src/main/java/com/sequenceiq/cloudbreak/client/CloudbreakIdentityClient.java [hunk 29]
```diff
@@ -167,12 +166,8 @@ public ClusterV1Endpoint clusterEndpoint() {
         return getEndpoint(ClusterV1Endpoint.class);
     }
 
-    public CredentialEndpoint credentialEndpoint() {
-        return getEndpoint(CredentialEndpoint.class);
-    }
-
-    public CredentialV3Endpoint credentialV3Endpoint() {
-        return getEndpoint(CredentialV3Endpoint.class);
+    public CredentialV4Endpoint credentialV3Endpoint() {
```
Comment (doktoric) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/client/CloudbreakIdentityClient.java]: credentialV4Endpoint

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Base.java [hunk 30]
```diff
@@ -0,0 +1,150 @@
...
+    @ApiModelProperty(value = ModelDescriptions.NAME, required = true, allowableValues = "length range[5, 100]")
+    private String name;
+
+    @NotNull
+    @ApiModelProperty(value = ModelDescriptions.CLOUD_PLATFORM, required = true)
+    private String cloudPlatform;
+
+    @Valid
+    @ApiModelProperty(CredentialModelDescription.AWS_PARAMETERS)
+    private AwsCredentialParameters aws;
```
Comment (doktoric) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/CredentialV4Base.java]: AwsCredentialV4Parameters

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/aws/KeyBasedCredentialParameters.java (hunk 26 above)
Comment (topolyai5) [reply core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/aws/KeyBasedCredentialParameters.java]: and allowableValues for every Enums

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/gcp/JsonParameters.java (hunk 14 above)
Comment (topolyai5) [reply core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/gcp/JsonParameters.java]: We have a Base64Deserializer and Base64Serializer class, I think it is usable here

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/yarn/YarnCredentialParameters.java [hunk 31]
```diff
@@ -0,0 +1,54 @@
...
+
+    public void setEndpoint(String endpoint) {
+        this.endpoint = endpoint;
+    }
+
+    @Override
+    public Map<String, Object> asMap() {
+        Map<String, Object> map = new LinkedHashMap<>();
+        map.put("cumulusAmbariUser", cumulusAmbariUser);
+        map.put("yarnEndpoint", endpoint);
```
Comment (topolyai5) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/credentials/parameters/yarn/YarnCredentialParameters.java]: it is different, is it ok?

**Pull request merged**